lib/grandpa/grandpa.go

```diff
@@ -679,6 +679,44 @@ func (s *Service) deleteVote(key ed25519.PublicKeyBytes, stage Subround) {
 	}
 }
 
+// implements `BeforeBestBlockBy` a custom voting rule that guarantees that our vote is always
+// behind the best block by at least N blocks, unless the base number is < N blocks behind the
+// best, in which case it votes for the base.
+// (https://github.com/paritytech/substrate/blob/1fd71c7845d6c28c532795ec79106d959dd1fe30/client/finality-grandpa/src/voting_rule.rs#L92)
+func (s *Service) determineBestHeaderToPrevote(finalizedHeader, bestBlockHeader *types.Header) (
+	headerToPrevote *types.Header, err error) {
+	gensisHash := s.blockState.GenesisHash()
+	isGenesisHash := gensisHash.Equal(bestBlockHeader.Hash())
+	if isGenesisHash || finalizedHeader.Hash().Equal(bestBlockHeader.Hash()) {
+		return bestBlockHeader, nil
+	}
+
+	isDescendant, err := s.blockState.IsDescendantOf(finalizedHeader.Hash(), bestBlockHeader.Hash())
+	if err != nil {
+		return headerToPrevote, fmt.Errorf("determine ancestry: %w", err)
```

Comment:
nit let's return explicitely `nil` to avoid any confusion 😉 

```suggestion
		return nil, fmt.Errorf("determining ancestry: %w", err)
```